File: backend/internal/objects/resultcards/cards.go
package resultcards

import (
	"time"

	"github.com/google/uuid"
)

// Card data types matching the frontend requirements
type CardData interface {
	GetType() string
}

type AuthorCardData struct {
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Timestamp   time.Time `json:"timestamp"`
	ExtraInfo   string    `json:"extraInfo,omitempty"`
	Index       int       `json:"index"`
	Type        string    `json:"type"`
	ObjectUUID  uuid.UUID `json:"object_uuid"`
}

func (a AuthorCardData) GetType() string {
	return "author"
}

type DocketCardData struct {
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Timestamp   time.Time `json:"timestamp"`
	Index       int       `json:"index"`
	Type        string    `json:"type"`
	ObjectUUID  uuid.UUID `json:"object_uuid"`
}

func (d DocketCardData) GetType() string {
	return "docket"
}

type DocumentAuthor struct {
	AuthorName      string    `json:"author_name"`
	IsPerson        bool      `json:"is_person"`
	IsPrimaryAuthor bool      `json:"is_primary_author"`
	AuthorID        uuid.UUID `json:"author_id"`
}

type DocumentConversation struct {
	ConvoName string    `json:"convo_name"`
	ConvoID   uuid.UUID `json:"convo_id"`
}

type DocumentCardData struct {
	Name           string               `json:"name"`
	Description    string               `json:"description"`
	Timestamp      time.Time            `json:"timestamp"`
	ExtraInfo      string               `json:"extraInfo,omitempty"`
	Index          int                  `json:"index"`
	Type           string               `json:"type"`
	ObjectUUID     uuid.UUID            `json:"object_uuid"`
	AttachmentUUID uuid.UUID            `json:"attachment_uuid"`
	FragmentID     string               `json:"fragment_id"`
	Authors        []DocumentAuthor     `json:"authors"`
	Conversation   DocumentConversation `json:"conversation"`
}

func (d DocumentCardData) GetType() string {
	return "document"
}

File: frontend/src/lib/requests/cards.ts
import axios from 'axios';
import { z } from "zod";
import { CardType } from "@/lib/types/generic_card_types";


export const BaseCardDataValidator = z.object({
  name: z.string(),
  object_uuid: z.string().uuid(),
  description: z.string(),
  timestamp: z.string(),
  extraInfo: z.string().optional(),
  index: z.number(),
});

export const AuthorCardDataValidator = BaseCardDataValidator.extend({
  type: z.literal(CardType.Author),
});

export const DocketCardDataValidator = BaseCardDataValidator.extend({
  type: z.literal(CardType.Docket),
});

export const DocumentCardDataValidator = BaseCardDataValidator.extend({
  type: z.literal(CardType.Document),
  authors: z.array(
    z.object({
      author_name: z.string(),
      is_person: z.boolean(),
      is_primary_author: z.boolean(),
      author_id: z.string(),
    })
  ),
});

export const CardDataValidator = z.union([
  AuthorCardDataValidator,
  DocketCardDataValidator,
  DocumentCardDataValidator,
]);

export type AuthorCardData = z.infer<typeof AuthorCardDataValidator>;
export type DocketCardData = z.infer<typeof DocketCardDataValidator>;
export type DocumentCardData = z.infer<typeof DocumentCardDataValidator>;
export type CardData = z.infer<typeof CardDataValidator>;

// API response validator
const SearchResponseValidator = z.object({
  results: z.array(CardDataValidator),
  total: z.number(),
  page: z.number(),
});

type SearchResponse = z.infer<typeof SearchResponseValidator>;

// API function with validation
export const searchCards = async (
  query: string,
  page: number = 1,
  limit: number = 10
): Promise<SearchResponse> => {
  try {
    const response = await axios.get('/api/search', {
      params: {
        q: query,
        page,
        limit
      }
    });

    // Validate the response data
    const validatedData = SearchResponseValidator.parse(response.data);
    return validatedData;
  } catch (error) {
    if (error instanceof z.ZodError) {
      console.error('Invalid response format:', error.errors);
      throw new Error('Invalid response format from server');
    }
    console.error('Search request failed:', error);
    throw error;
  }
};

// Utility function to format timestamp
const formatTimestamp = (timestamp: string): string => {
  return new Date(timestamp).toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
    hour: '2-digit',
    minute: '2-digit'
  });
};
